Add Webp support

sneako opened the pull request:
Hi and thanks for the lib!

I needed webp support, so I ported the Ruby implementation from: https://github.com/sdsykes/fastimage/blob/master/lib/fastimage.rb#L591

Let me know what you think

sneako committed:
Add webp support, with implementation ported for FastImage Ruby gem

sneako committed:
refactor tests

sneako committed:
use module attrs in test instead of redeclaring dimensions

sneako committed:
fix credo warnings

sneako commented on test/fastimage_test.exs:
```diff
@@ -11,16 +15,25 @@ defmodule FastimageTest do
   @bmp_url "https://raw.githubusercontent.com/stephenmoloney/fastimage/master/priv/test.bmp"
+  @webp_vp8l_url "https://raw.githubusercontent.com/sdsykes/fastimage/master/test/fixtures/webp_vp8l.webp"
+  @webp_vp8_url "https://raw.githubusercontent.com/sdsykes/fastimage/master/test/fixtures/webp_vp8.webp"
+  @webp_vp8x_url "https://raw.githubusercontent.com/sdsykes/fastimage/master/test/fixtures/webp_vp8x.webp"
```
we can point these urls at this repo, but left them as is so you could see the tests pass first

stephenmoloney replied:
Let's use `https://raw.githubusercontent.com/stephenmoloney/fastimage/master/priv/webp_*.webp` instead.

stephenmoloney replied:
and thanks for pointing that out BTW.

sneako committed:
add local webp benchmarks

stephenmoloney commented:
Hey @sneako, thanks for the PR. Webp support is a good addition.
I don't use bitwise operators much so I'll rely on you for that part.

stephenmoloney commented on bench/fastimage_bench.exs:
```diff
@@ -10,6 +10,9 @@ defmodule FastimageBench do
   @png_file "./priv/test.png"
   @gif_file "./priv/test.gif"
   @bmp_file "./priv/test.bmp"
+  @webp_vp8_file "./priv/test_vp8.webp"
```
Are these images (./priv/test_*.webp) taken by you and owned by you ? [just to avoid any potential copyright issue]

sneako replied:
These images were all taken from the test suite of the Ruby gem FastImage, which also has an MIT license https://github.com/sdsykes/fastimage/blob/master/MIT-LICENSE
So we should be ok to use them as well, right?

stephenmoloney replied:
I found that image on [Google's webp site](https://developers.google.com/speed/webp/gallery1#image_credits), seems they got it from somewhere else after seeking permission and giving credit which is the correct etiquette for this project too if we use it (which is a possibility) although for reasons mentioned later, I'd prefer to recreate 3 new images of `priv/test.jpeg` with the webp formats in the parser.

stephenmoloney commented on test/fastimage_test.exs:
```diff
@@ -192,32 +164,46 @@ defmodule FastimageTest do
+    assert actual_size == {:ok, expected_size}
+  end
+
+  defp list(n) do
```
Nice addition

stephenmoloney commented on test/fastimage_test.exs:
```diff
@@ -3,6 +3,10 @@ defmodule FastimageTest do
   doctest Fastimage
 
   @expected_size %Fastimage.Dimensions{width: 283, height: 142}
+  @webp_vp8_expected_size %Fastimage.Dimensions{width: 550, height: 368}
```
Now that there are multiple sizes, `@expected_size` on it's own is a little bit more confusing. Not sure of what alternative though.

sneako replied:
I know what you mean. Maybe something like `standard_expected_size`? It's just a tad bit less ambiguous

stephenmoloney replied:
What might be cleaner is to just create the `test_vp*.webp` images using the same one as already in this lib and then fix it to `@expected_size` for consistency and drop the external image from sdykes lib. I just need to figure out how to create the three different webp formats.

sneako replied:
I used https://developers.google.com/speed/webp/docs/cwebp to resize the images to match the `expected_size`. I made sure that the tool did not change the type of webp (vp8*) of each new file.

sneako commented:
@stephenmoloney glad you like the feature. Honestly, I do not use bitwise operators much either, but it was easy to translate from the ruby code and passes the tests. I'm surprised to see test coverage has decreased, when i run the tests on master, it reports 61.4%, so I thought coverage increased. I'll update the urls to point at this repo now. Is there anything else you would like me to change before merging?

stephenmoloney reviewed (changes requested):
I think this looks good to merge.

The main thing is to clarify 

- You are the owner of the images and are releasing them under the project `MIT` license 
- Change the http url for the images as discussed in the review

Thanks !

sneako committed:
switch webp test img urls

stephenmoloney commented:
Also, need to add a small mention of support for webp [here](https://github.com/stephenmoloney/fastimage#features)

sneako committed:
resize webp test images

sneako committed:
update docs

sneako commented:
@stephenmoloney I noticed there were actually several places that had the full list of supported types, and the options docs were repeated too so I did some consolidation. What do you think?

stephenmoloney commented on lib/fastimage.ex:
```diff
@@ -49,14 +29,38 @@ defmodule Fastimage do
+  @type fastimage_opts :: [
+          stream_timeout: non_neg_integer,
+          max_error_retries: non_neg_integer,
+          max_error_retries: non_neg_integer
```
meant to be `max_redirect_retries`

stephenmoloney replied:
I fixed this too.

stephenmoloney commented on test/fastimage_test.exs:
```diff
@@ -3,24 +3,35 @@ defmodule FastimageTest do
+  @webp_vp8x_binary File.read!(@webp_vp8x_file)
+  @webp_vp8l_binary File.read!(@webp_vp8l_file)
 
   @tag :jpeg
```
Not sure why I put this tag here, I think we can remove this and use the same `assert_size_and_type/3` function inside this test for consistency

stephenmoloney replied:
I fixed that.

Commit:
Add webp images and test

What does this commit/MR/PR do?

- Add webp images
- Test webp images
- Rearrange few test functions

Why is this commit/MR/PR needed?

- To test webp without external image dependencies

stephenmoloney commented:
@sneako 

I have made a few changes to your branch.

Would you mind merging this branch `feature/webp` into your branch in this PR ?

Ref https://github.com/sneako/fastimage/pull/1

Thanks !

Commit:
Add similar changes for the benchmarks

What does this commit/MR/PR do?

- Add similar changes for the benchmarks

Why is this commit/MR/PR needed?

- So that benchmarks are also covered by changes

sneako committed:
Merge pull request #1 from stephenmoloney/feature/webp

Feature/webp

sneako commented:
Thanks for taking care of those updates!

stephenmoloney commented:
:+1: :tada:
